### ISSUE TITLE
[$1000] The cursor resets to the beginning for the edit message in draft

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#21137]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to any report
2. Send any comment
3. Long press the comment
4. Press Edit comment
5. Set cursor position anywhere except at the beginning
6. Go to any other report
7. Come back to the report that has edit draft
## Expected Result: 
The cursor position remains unchanged
## Actual Result: 
The cursor position resets to the beginning

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.29-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/ca8641a6-4984-4242-ac0e-3f033c691473

https://github.com/Expensify/App/assets/43996225/254a5d80-ac77-4d63-989e-433d274b30fe

**Expensify/Expensify Issue URL:**
**Issue reported by:** @adeel0202 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1686597604159949

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~014a2eb8967fd1486f</li>
        <li>Upwork Job ID: 1671935895187853312</li>
        <li>Last Price Increase: 2023-06-29</li>
    </ul>
</details>

**Comment by slafortune** [comment]
Looks good!

**Comment by Thanos30** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The selection (cursor position / text selected) of the draft edit message is not persistent when we exit and re-enter a chat report.

### What is the root cause of that problem?

We have this code here on `ReportActionItemMessageEdit.js`:

```
/**
     * Save the draft of the comment. This debounced so that we're not ceaselessly saving your edit. Saving the draft
     * allows one to navigate somewhere else and come back to the comment and still have it in edit mode.
     * @param {String} newDraft
     */
    const debouncedSaveDraft = useMemo(
        () =>
            _.debounce((newDraft) => {
                Report.saveReportActionDraft(props.reportID, props.action.reportActionID, newDraft);
            }, 1000),
        [props.reportID, props.action.reportActionID],
    );

```
The `saveReportActionDraft()` function is like that:

```
function saveReportActionDraft(reportID, reportActionID, draftMessage) {
    Onyx.set(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS_DRAFTS}${reportID}_${reportActionID}`, draftMessage);
}
```

As you can see, we are only saving the `draftMessage` on Onyx, and not the values of the `start` and `end` of the selection.

### What changes do you think we should make in order to solve the problem?
If we want to keep the `selection` value, we should also save the selection on Onyx as we do with the `draftMessage`.

This will also demand to make changes on all the different components that share the `draftMessage`, I will avoid adding all of them not to make this proposal too long, but I can provide a list of changed files below.
In general we should save the `draftMessage` like this:

```
function saveReportActionDraft(reportID, reportActionID, draftMessage, selection) {
    Onyx.set(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS_DRAFTS}${reportID}_${reportActionID}`, {message: draftMessage, selection});
}

```

**List of changed files:**

- Report.js
- ReportActionItem.js
- ReportActionItemMessageEdit.js
- ContextMenuActions.js
- genericReportActionContextMenuPropTypes.js
- PopoverReportActionContextMenu.js


PS. This happens on all devices.

**Comment by kameshwarnayak** [comment]
### Proposal
### Please re-state the problem that we are trying to solve in this issue.
The cursor resets to the beginning for the edit message in draft

### What is the root cause of that problem?
The selection values in the react native TextInput are not being set in few scenarios. There are multiple open bugs relating TextInput selection option ([29063](https://github.com/facebook/react-native/issues/29063), [35005](https://github.com/facebook/react-native/issues/35005)) in react native GH. However they don't point to the exact scenario that we are facing. 

One of the workaround mentioned is to use setNativeProps to set the selection values. 

### What changes do you think we should make in order to solve the problem?

Using setNativeProps to set the selection in the use effects of the following file will solve the issue

https://github.com/Expensify/App/blob/820fa93ad1e799c12d8b639b98018b5350ea17af/src/pages/home/report/ReportActionItemMessageEdit.js#L98-L101

becomes

```
    useEffect(() => {
        // required for keeping last state of isFocused variable
        isFocusedRef.current = isFocused;
        textInputRef.current.setNativeProps({
            selection: selection
        });
    }, [isFocused]);
```

Result :

https://github.com/Expensify/App/assets/2232432/280a0f9e-442e-4cb5-a1e6-a4d07dfa4230



### What alternative solutions did you explore? (Optional)
None

**Comment by slafortune** [comment]
@thesahindia - thoughts?

**Comment by kameshwarnayak** [comment]
> @kameshwarnayak's [proposal](https://github.com/Expensify/App/issues/21137#issuecomment-1605507856) looks good.
> 
> 🎀 👀 🎀 C+ reviewed

@thesahindia Thanks for the confirmation. I have submitted the proposal on Upwork. Will create the PR once I get the approval on Upwork.

**Comment by aldo-expensify** [comment]
I'll be out of office next week, I don't think you need any action from me 👋

**Comment by aldo-expensify** [comment]
PR merged

**Comment by slafortune** [comment]
@thesahindia can you please complete the checklist?

**Comment by Santhosh-Sellavel** [comment]
@aldo-expensify @slafortune 

Sorry for the late notice. The PR has caused a [regression](https://github.com/Expensify/App/pull/22443/files#r1271101903), please hold on for any action here, thanks!

**Comment by Santhosh-Sellavel** [comment]
This is the regression. Now this can be reproduced on main.

https://github.com/Expensify/App/assets/85645967/714fd14e-8981-4c77-b9d3-10741ad73212

**Comment by joekaufmanexpensify** [comment]
As an fyi, the PR in this issue caused another regression[ here](https://github.com/Expensify/App/issues/23451#issuecomment-1648682606).

**Comment by aldo-expensify** [comment]
I think we should do a revert and then check with more time what is a better way to fix this

**Comment by aldo-expensify** [comment]
Revert: https://github.com/Expensify/App/pull/23518

**Comment by aldo-expensify** [comment]
> I tried to recreate the issue with the latest main, and I can't reproduce it. Can anyone please confirm if we care recreate this?

Are you referring to this issue: https://github.com/Expensify/App/issues/21137#issuecomment-1646963074?

I reproduced it, and I confirmed the revert fixes it: https://github.com/Expensify/App/pull/23518
Maybe you can't reproduce because the revert was already merged to main?

**Comment by daordonez11** [comment]
@kameshwarnayak  I'm able to reproduce it's just that instead of resetting to the beginning for me now it's resetting to the end instead of keeping the position.

**Comment by slafortune** [comment]
@kameshwarnayak any update on this?

**Comment by slafortune** [comment]
I'm having a hard time understanding what is complete in what timeframe on this GH and trying to set up payment and reassign this. @thesahindia are you able to summarize this for us?

CC @aldo-expensify  @kameshwarnayak

**Comment by slafortune** [comment]
reassigning a BZ member, I am out on sabbatical

**Comment by bfitzexpensify** [comment]
@thesahindia bump on this question, so I can get payments sorted out:

>I'm having a hard time understanding what is complete in what timeframe on this GH and trying to set up payment and reassign this. @thesahindia are you able to summarize this for us?

**Comment by bfitzexpensify** [comment]
Sure, title updated. 

So, my understanding is that we're waiting on the 2nd PR to be reviewed, merged and deployed, and then we can follow the typical close-out process (BZ checklist and payments issued) after that's done — is that correct?

**Comment by bfitzexpensify** [comment]
Awesome, and https://github.com/Expensify/App/pull/24552 is the new PR, right?

**Comment by kameshwarnayak** [comment]
@thesahindia Was a bit held up with a personal commitment. I will have a look at the comment in a couple of days.

**Comment by kameshwarnayak** [comment]
Have we changed the behaviour of the focus on returning to a report. The focus is not on the draft when we open a report in the current main. Is it intended or a regression? 

https://github.com/Expensify/App/assets/2232432/f62595bc-ad06-424b-851c-7495dbcda602

**Comment by aldo-expensify** [comment]
cc @Beamanator since you were talking the other day about documenting this kind of things (like focus behaviour) ^

> The focus is not on the draft when we open a report in the current main. Is it intended or a regression?

No idea. For me it makes sense that it should be focused automatically, but I don't know if someone thought differently and removed it on purpose or if it is a regression.

**Comment by kameshwarnayak** [comment]
@aldo-expensify My personal choice would be to focus automatically. 

Should we proceed with that change or find if the change is on purpose? If there is a document with these changes, it would be great to refer.

**Comment by bfitzexpensify** [comment]
OK, I think it might be related to this broader issue - https://github.com/Expensify/App/issues/15992 - that's dealing with general Composer Component Focus Issues. Does that sound right to everyone else?

**Comment by aldo-expensify** [comment]
I think it would make a lot of sense to define the focus behaviour we want and write it down so we don't end up changing it after because of someone personal preference.

**Comment by aldo-expensify** [comment]
I'm going OOO today, please reassign if you think you need an engineer. I'm not sure about where we are with this issue... it has the label "Awaiting Payment"

**Comment by bfitzexpensify** [comment]
I'm not entirely sure, either. @thesahindia, what are your thoughts on where this is at? https://github.com/Expensify/App/pull/24552 was opened, but not merged, and my understanding is that the intent was to finalise that and then action payment on that PR. 

https://github.com/Expensify/App/issues/15992 is open, set to monthly, and unassigned, so I don't think anything will happen there anytime soon. 

So, should we move forward with https://github.com/Expensify/App/pull/24552, and then wrap this up, or should we leave this entirely?

**Comment by kameshwarnayak** [comment]
@bfitzexpensify https://github.com/Expensify/App/pull/24552 this will not work because the behaviour has changed. With the current behaviour, we may not need a fix. 

I am confused on the behaviour of the app when the user goes back to a report with a draft. Let me check the latest main and get back on that.

**Comment by bfitzexpensify** [comment]
Any update on this one @kameshwarnayak?

**Comment by bfitzexpensify** [comment]
Friendly bump @kameshwarnayak

**Comment by kameshwarnayak** [comment]
@bfitzexpensify - will look into it today. Was but held up and but under the weather

**Comment by kameshwarnayak** [comment]
@bfitzexpensify Tested this on the latest master. The behaviour has changed and the keyboard doesn't appear by default. Please find the video below. 

https://github.com/Expensify/App/assets/2232432/7dc6521f-1ac6-4615-90df-c4066a9c4d85

If this is the new behaviour, then the bug becomes irrelevant.

**Comment by bfitzexpensify** [comment]
Agreed @kameshwarnayak. Closing this out.
